tests/suites/test_suite_psa_crypto_init.function

```diff
@@ -112,6 +113,65 @@ static void custom_entropy_init(mbedtls_entropy_context *ctx)
 
 #endif /* !defined(MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG) */
 
+#if defined MBEDTLS_THREADING_PTHREAD
+
+typedef struct {
+    int do_shutdown;
+} thread_psa_init_ctx_t;
+
+static void *thread_psa_init_function(void *ctx)
+{
+    thread_psa_init_ctx_t *init_context = (thread_psa_init_ctx_t *) ctx;
+    psa_status_t status = PSA_ERROR_CORRUPTION_DETECTED;
+    uint8_t random[10] = { 0 };
+
+    PSA_ASSERT(psa_crypto_init());
+
+    if (init_context->do_shutdown) {
+        /* Note that shutting down PSA crypto is *not* guaranteed to be thread
+         * safe if *any* operations are currently in progress. I am merely doing
+         * this here to test the safety of global flags that are written in
+         * psa_crypto_init() / mbedtls_psa_crypto_free(). */
+        mbedtls_psa_crypto_free();
+    }
+
+    /* Given that another thread may have called mbedtls_psa_crypto_free() by
+     * this point on another thread, we can only guarantee the return codes
+     * from the following tests if we are not also testing shutdown. The tests
+     * are still worth doing though, so that TSAN can catch potential issues. */
+
+    /* Test getting if drivers are initialised. */
+    int can_do = psa_can_do_hash(PSA_ALG_NONE);
+
+    if (!init_context->do_shutdown) {
+        TEST_ASSERT(can_do == 1);
+    }
+
+#if !defined(MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG)
+
+    /* Test getting global_data.rng_state. */
+    status = mbedtls_psa_crypto_configure_entropy_sources(NULL, NULL);
+
+    if (!init_context->do_shutdown) {
+        /* Bad state due to entropy sources already being setup in
+         * psa_crypto_init() */
+        TEST_EQUAL(status, PSA_ERROR_BAD_STATE);
+    }
+#endif /* !defined(MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG) */
+
+    /* Test using the PSA RNG if not testing shutdown (RNG is not safe to use if
+     * we shut down PSA whilst using it). */
+    if (!init_context->do_shutdown) {
+        status = psa_generate_random(random, sizeof(random));
+
+        TEST_EQUAL(status, PSA_SUCCESS);
+    }
```

Comment:
Suggestion: We can shorten this by explicitely splitting the cases,  but if you think the current style is cleaner then keep it as is
``` 
    if (init_context->do_shutdown) {
        /* Note that shutting down PSA crypto is *not* guaranteed to be thread
         * safe if *any* operations are currently in progress. I am merely doing
         * this here to test the safety of global flags that are written in
         * psa_crypto_init() / mbedtls_psa_crypto_free(). */
        mbedtls_psa_crypto_free();
        goto exit;
    } else {

        /* Test getting if drivers are initialized.
         * We cannot test this if do_shutdown is true, as another thread
         * may have already called mbedtls_psa_crypto_free
         * which uninitializes the drivers. TSAN can catch potential issues
         * here, so we should still do this test. */
        TEST_ASSERT(psa_can_do_hash(PSA_ALG_NONE) == 1);

#if !defined(MBEDTLS_PSA_CRYPTO_EXTERNAL_RNG)
        ....
#endif
        ....
    }
```

Reply:
This has kinda changed in the latest iteration. I was, however deliberately running the flag test checks at the same time as doing startup / shutdown of PSA.